plugins/gpio/fu-plugin-gpio.c

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (C) 2022 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <fwupdplugin.h>
+
+#include "fu-gpio-device.h"
+
+struct FuPluginData {
+	FuGpioDevice *gpio_device;
+};
+
+static void
+fu_plugin_gpio_init(FuPlugin *plugin)
+{
+	FuContext *ctx = fu_plugin_get_context(plugin);
+	fu_plugin_alloc_data(plugin, sizeof(FuPluginData));
+	fu_context_add_quirk_key(ctx, "GpioForUpdate");
+	fu_plugin_add_udev_subsystem(plugin, "gpio");
+	fu_plugin_add_device_gtype(plugin, FU_TYPE_GPIO_DEVICE);
+}
+
+static void
+fu_plugin_gpio_destroy(FuPlugin *plugin)
+{
+	FuPluginData *data = fu_plugin_get_data(plugin);
+	if (data->gpio_device != NULL)
+		g_object_unref(data->gpio_device);
+}
+
+static gboolean
+fu_plugin_gpio_parse_level(const gchar *str, gboolean *ret, GError **error)
+{
+	if (g_strcmp0(str, "high") == 0) {
+		*ret = TRUE;
+		return TRUE;
+	}
+	if (g_strcmp0(str, "low") == 0) {
+		*ret = FALSE;
+		return TRUE;
+	}
+	g_set_error(error,
+		    G_IO_ERROR,
+		    G_IO_ERROR_INVALID_DATA,
+		    "cannot parse level, got %s and expected high|low",
+		    str);
+	return FALSE;
+}
+
+static gboolean
+fu_plugin_gpio_process_quirk(FuPlugin *self, const gchar *str, GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data(self);
+	gboolean value = FALSE;
+	FuDevice *device;
+	g_auto(GStrv) split = g_strsplit(str, ",", -1);
+	g_autoptr(FuDeviceLocker) locker = NULL;
+
+	/* already set by previous GUID */
+	if (data->gpio_device != NULL)
```

Comment:
couldn't this also happen if multiple GPIO devices came by?  In which case you have a probe order race condition.